source/fuzz/fuzzer_pass_permute_function_variables.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2021 Mostafa Ashraf
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_permute_function_variables.h"
+
+#include <algorithm>
+#include <numeric>
+#include <vector>
+
+#include "source/fuzz/fuzzer_context.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassPermuteFunctionVariables::FuzzerPassPermuteFunctionVariables(
+    opt::IRContext* ir_context, TransformationContext* transformation_context,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}  // Here we call parent constructor
+
+void FuzzerPassPermuteFunctionVariables::Apply() {
+  // Permuting OpVariable instructions in each function
+  for (auto& function : *GetIRContext()->module()) {
+    if (!GetFuzzerContext()->ChoosePercentage(
+            GetFuzzerContext()->GetChanceOfPermuteFunctionVariables())) {
+      continue;
+    }
+    // Ids storage section
+    auto first_block = function->entry().get();
+
+    std::vector<opt::Instruction*> variables;
+    for (auto& instruction : first_block) {
+      if (instruction.opcode() == SpvOpVariable) {
+        variables.push_back(&instruction);
+      }
+    }
+
+    uint32_t vars_size = variables.size();
+
+    // Permutation section
+    // Create a vector, fill it with [0, n-1] values and shuffle it
+    std::vector<uint32_t> permutation(vars_size);
+    std::iota(permutation.begin(), permutation.end(), 0);
+    GetFuzzerContext()->Shuffle(&permutation);
+
+    std::vector<std::pair<uint32_t, uint32_t>> variables_pair_id;
+    /*
+    // Cycle notation, Apply product of transpositions because
+    // Every Permutation can be written as a product of transpositions and
+    // Transpositions Is special case of Permutation but between two numbers.
+    // Mathematical formula I've followed
+    // (a1,a2,...,as)=(as,as−1)∘(as,as−2)∘...∘(as,a2)∘(as,a1)
+    */
+    for (uint32_t changed_index = vars_size - 2; changed_index > 0;
+         changed_index--) {
+      variables_pair_id.push_back(std::make_pair(vars_size - 1, changed_index));
+    }
```

Comment:
Thanks, I've followed this approach I think it enhances the complexity.

Reply:
Having a complicated implementation increases the possibility of introducing bugs. For example:
- The part, where you are applying the product of transpositions algorithm, doesn't take into account the fact that the function might have less than 2 variables. If this happens, the assignment to `changed_index` will overflow (since you are using unsigned integers). When the overflow happens, you'll get an undefined behaviour when you try to swap variables with some huge indices.
- `changed_index > 0` is a tricky one - it doesn't take into account the variable at 0 index. Thus, it will never be swapped with anything. At the same time, if you replace the comparison with `changed_index >= 0`, you'll get an overflow while decrementing 1 from the `changed_index` on the last iteration of the loop. Simply put, `changed_index >= 0` always true for unsigned integers. Thus, you get an infinite loop.
- The implementation of the product of transpositions isn't correct. Consider the case when the `shuffle` method produces the original sequence (i.e. no permutation is happening). The product of transpositions is supposed to produce the desired permutation when applied to a sequence of elements. In this case, however, it will only move the first element to the last position. To illustrate:

If you had this sequence of elements:
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

And the `shuffle` produced the following permutation:
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Then the product of transpositions is supposed to produce an original sequence (`[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`) when applied to `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`. However, the current implementation will produce
> [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]

Feel free to correct me, though, if you think I'm wrong :)

The fact that you're aware of the 'product of transpositions' algorithm clearly shows your proficiency with algorithms. However, in this case, it is easier and less buggy to have the implementation I described in the previous comment. The other reason why a simpler implementation is better is the fact that it will be harder for people to understand the code if they are not familiar with the `product of transpositions` algorithm.

With all this in mind, I would ask you to change the implementation to the one described in the previous comment.

Reply:
By the way, I just spoke with Ally and turns out he was the one who asked you to do the product of transpositions approach :) We discussed it a bit and agreed that it would indeed be easier to use the approach I described in https://github.com/KhronosGroup/SPIRV-Tools/pull/4213#discussion_r606815928 instead. Sorry for the confusion :)

Reply:
Regarding the fact that you are swapping variables based on their indices, it is better to use result ids instead. The reason is better compatibility with the shrinker option of the spirv-fuzz. The shrinker accepts a buggy SPIR-V and tries to remove as many applied transformations as possible as long as the bug still occurs. With this in mind, consider the following scenario.

Suppose we have 2 transformations:
- `AddLocalVariable(i)` - adds a new local variable to the function. The newly added variable will have an index `i`.
- `SwapVariables(i, j)` - swaps 2 variables given their indices in the function.

Suppose that we have sequence of transformations:
- `AddLocalVariable(i)`
- `SwapVariables(j, k)`

Also let's further assume that `SwapVariables(j, k)` triggers some bug. In this case, if `j > i || k > i`, then when the shrinker removes the first transformation, the second one will be swapping different variables and thus may not cause the bug at all.

All this might be somewhat complicated and you don't need to know the internals of how the shrinker works right now. But the idea is that we are trying to design our transformations such that they either apply some specific change (e.g. permute specific variables with result ids `a` and `b`) or none at all.

Let me know if you need any help/clarification :)

Reply:
Thanks Eng. @Vasniktel

Reply:
I've implemented this, I think it will work in all cases and not fail because a `RemoveAtRandomIndex ` returns the element at a specific index then remove it from the vector so if the vector is even size will work fine, if odd will ignore the last element. what do you think @Vasniktel, @afd?
```c++ 
for (; variables.size() >= 2;) {
      auto instruction_1 = GetFuzzerContext()->RemoveAtRandomIndex(&variables);
      auto instruction_2 = GetFuzzerContext()->RemoveAtRandomIndex(&variables);

      ApplyTransformation(TransformationSwapFunctionVariables(
          instruction_1->result_id(), instruction_2->result_id()));
    }
```
Thanks.

Reply:
Yes, that looks good to me. One minor remark is you can replace the `for` loop with the `while`.

Let me know when you push the changes and ready for a review.

Reply:
@Vasniktel  I've pushed my last updates